File: core/psg.c
/*
 * psg.c — Sega PSG (SN76489/SN76496 variant used in Master System / Genesis)
 *
 * Faithful register-level model, not a frequency-domain shortcut:
 *   - 3 tone channels, each a 10-bit down-counter driving a toggle flip-flop
 *   - 1 noise channel: same counter mechanism, feeding a 16-bit LFSR that
 *     shifts only on the counter flip-flop's 0->1 transitions
 *   - White noise feedback taps bits 0 and 3 (XOR/parity), fed back into
 *     bit 15; periodic noise taps only bit 0. LFSR resets to 0x8000 on any
 *     write to the noise register. (Sega variant: 16-bit register, not the
 *     original TI 15-bit one.)
 *   - 4-bit attenuation per channel, -2dB per step, matching the widely
 *     published SN76489 volume table.
 * Reference: SMS Power! development wiki, "SN76489" page.
 */
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

typedef struct {
    uint16_t tone_reg[3];
    uint8_t  vol_reg[4];
    uint8_t  noise_reg;

    int32_t  tone_counter[3];
    uint8_t  tone_output[3];

    int32_t  noise_counter;
    uint8_t  noise_ff;
    uint16_t lfsr;

    uint8_t  latched_channel;
    uint8_t  latched_type;
} PSGHandle;

static const int16_t VOLUME_TABLE[16] = {
    32767, 26028, 20675, 16422, 13045, 10362, 8231, 6568,
    5193,  4125,  3277,  2603,  2067,  1642,  1304, 0
};

PSGHandle *psg_new(void) {
    PSGHandle *p = (PSGHandle *)calloc(1, sizeof(PSGHandle));
    if (!p) return NULL;
    p->vol_reg[0] = p->vol_reg[1] = p->vol_reg[2] = p->vol_reg[3] = 0x0F;
    p->tone_output[0] = p->tone_output[1] = p->tone_output[2] = 1;
    p->lfsr = 0x8000;
    return p;
}

void psg_free(PSGHandle *p) { free(p); }

void psg_reset(PSGHandle *p) {
    memset(p, 0, sizeof(PSGHandle));
    p->vol_reg[0] = p->vol_reg[1] = p->vol_reg[2] = p->vol_reg[3] = 0x0F;
    p->tone_output[0] = p->tone_output[1] = p->tone_output[2] = 1;
    p->lfsr = 0x8000;
}

void psg_write(PSGHandle *p, unsigned char byte) {
    if (byte & 0x80) {
        uint8_t channel = (byte >> 5) & 0x03;
        uint8_t type = (byte >> 4) & 0x01;
        uint8_t data = byte & 0x0F;
        p->latched_channel = channel;
        p->latched_type = type;
        if (type) {
            p->vol_reg[channel] = data;
        } else if (channel == 3) {
            p->noise_reg = data & 0x07;
            p->lfsr = 0x8000;
        } else {
            p->tone_reg[channel] = (p->tone_reg[channel] & 0x3F0) | data;
        }
    } else {
        uint8_t data6 = byte & 0x3F;
        if (p->latched_type) {
            p->vol_reg[p->latched_channel] = data6 & 0x0F;
        } else if (p->latched_channel == 3) {
            p->noise_reg = data6 & 0x07;
            p->lfsr = 0x8000;
        } else {
            p->tone_reg[p->latched_channel] =
                (uint16_t)((p->tone_reg[p->latched_channel] & 0x00F) | (data6 << 4));
        }
    }
}

static int parity16(uint16_t v) {
    v ^= v >> 8; v ^= v >> 4; v ^= v >> 2; v ^= v >> 1;
    return v & 1;
}

static void psg_tick(PSGHandle *p) {
    int ch;
    for (ch = 0; ch < 3; ch++) {
        if (p->tone_reg[ch] <= 1) {
            p->tone_output[ch] = 1; /* DC output, per spec (sample-playback trick) */
            continue;
        }
        /* Reload-then-decrement, NOT reload-or-decrement. Testing first and
         * decrementing only in the else branch spends one extra tick per
         * half-period, which makes every tone channel play at
         * clock/(32*(N+1)) instead of the clock/(32*N) that freq_to_tone_n
         * assumes -- about 7 cents flat, consistent across the register
         * range, so it reads as "the PSG is slightly out of tune with the
         * FM" rather than as a bug. */
        if (p->tone_counter[ch] <= 0) {
            p->tone_counter[ch] = p->tone_reg[ch];
            p->tone_output[ch] ^= 1;
        }
        p->tone_counter[ch]--;
    }

    int reset_val;
    switch (p->noise_reg & 0x03) {
        case 0: reset_val = 0x10; break;
        case 1: reset_val = 0x20; break;
        case 2: reset_val = 0x40; break;
        default: reset_val = p->tone_reg[2] > 0 ? p->tone_reg[2] : 1; break;
    }
    if (p->noise_counter <= 0) {
        p->noise_counter = reset_val;   /* same reload-then-decrement rule */
        uint8_t old_ff = p->noise_ff;
        p->noise_ff ^= 1;
        if (old_ff == 0 && p->noise_ff == 1) {
            int white = (p->noise_reg & 0x04) != 0;
            uint16_t fb = white ? (uint16_t)parity16(p->lfsr & 0x0009)
                                 : (uint16_t)(p->lfsr & 1);
            p->lfsr = (uint16_t)((p->lfsr >> 1) | (fb << 15));
        }
    }
    p->noise_counter--;
}

/* Render n_samples of native-rate (chip_clock/16) mono audio, int16. */
void psg_render(PSGHandle *p, int n_samples, short *out) {
    int i, ch;
    for (i = 0; i < n_samples; i++) {
        psg_tick(p);
        int32_t mix = 0;
        for (ch = 0; ch < 3; ch++) {
            int bit = p->tone_output[ch] ? 1 : -1;
            mix += bit * VOLUME_TABLE[p->vol_reg[ch]];
        }
        int nbit = (p->lfsr & 1) ? 1 : -1;
        mix += nbit * VOLUME_TABLE[p->vol_reg[3]];
        mix /= 4;
        if (mix > 32767) mix = 32767;
        if (mix < -32768) mix = -32768;
        out[i] = (short)mix;
    }
}

File: python/fallback/psg.py
"""
fallback/psg.py — the Sega PSG in pure Python.

Same model as core/psg.c: three 10-bit down-counters driving toggle
flip-flops, one 16-bit LFSR with taps on bits 0 and 3 for white noise and
bit 0 alone for periodic, the published 4-bit / -2 dB attenuation table.

The difference is arithmetic, not semantics. psg.c steps every one of the
chip's 223,721 ticks per second because in C that is free. Doing that in
Python would cost roughly a second of CPU per second of audio, so instead
each output sample integrates the square wave over the ticks it spans:

    a tone channel toggles every N ticks, so over a window of T ticks the
    output is high for a computable fraction of the time

which is both faster (a handful of toggles per sample rather than five
ticks x four channels) and better behaved, because integrating over the
window is a box filter — the hard edges alias far less than point-sampling
them would. The register semantics, the LFSR, and the volume table are
identical to the C core.
"""

from array import array

import audio as _audio

NTSC_PSG_CLOCK = 3_579_545

#: 4-bit attenuator, -2 dB per step, 15 = silence. Same table as psg.c.
VOLUME_TABLE = (32767, 26028, 20675, 16422, 13045, 10362, 8231, 6568,
                5193, 4125, 3277, 2603, 2067, 1642, 1304, 0)

#: Noise counter reload for modes 0-2; mode 3 follows tone channel 2.
NOISE_RELOAD = (0x10, 0x20, 0x40)


class PyPSG:
    """Drop-in for the ctypes SN76489 handle used by sn76489.SN76489."""

    def __init__(self, clock: float = NTSC_PSG_CLOCK):
        self.clock = float(clock)
        self.rate = self.clock / 16.0
        self.reset()

    def reset(self):
        self.tone_reg = [0, 0, 0]
        self.vol_reg = [0x0F] * 4
        self.noise_reg = 0
        self.tone_phase = [0.0, 0.0, 0.0]   # ticks until the next toggle
        self.tone_output = [1, 1, 1]
        self.noise_phase = 0.0
        self.noise_ff = 0
        self.lfsr = 0x8000
        self._latched_channel = 0
        self._latched_type = 0

    # -- bus ---------------------------------------------------------------
    def write(self, byte: int):
        byte &= 0xFF
        if byte & 0x80:
            channel = (byte >> 5) & 0x03
            type_ = (byte >> 4) & 0x01
            data = byte & 0x0F
            self._latched_channel = channel
            self._latched_type = type_
            if type_:
                self.vol_reg[channel] = data
            elif channel == 3:
                self.noise_reg = data & 0x07
                self.lfsr = 0x8000
            else:
                self.tone_reg[channel] = (self.tone_reg[channel] & 0x3F0) | data
        else:
            data6 = byte & 0x3F
            if self._latched_type:
                self.vol_reg[self._latched_channel] = data6 & 0x0F
            elif self._latched_channel == 3:
                self.noise_reg = data6 & 0x07
                self.lfsr = 0x8000
            else:
                channel = self._latched_channel
                self.tone_reg[channel] = (self.tone_reg[channel] & 0x00F) | (data6 << 4)

    # -- LFSR --------------------------------------------------------------
    def _shift_lfsr(self):
        if self.noise_reg & 0x04:                 # white: parity of bits 0 and 3
            v = self.lfsr & 0x0009
            v ^= v >> 8
            v ^= v >> 4
            v ^= v >> 2
            v ^= v >> 1
            feedback = v & 1
        else:                                     # periodic: bit 0 straight back
            feedback = self.lfsr & 1
        self.lfsr = ((self.lfsr >> 1) | (feedback << 15)) & 0xFFFF

    # -- rendering ---------------------------------------------------------
    def render(self, n_samples: int):
        """n_samples of mono float audio at self.rate."""
        if n_samples <= 0:
            return _audio.zeros(0, 1)

        data = array("f", bytes(4 * n_samples))

        ticks_per_sample = 1.0    # this core's rate IS the chip tick rate
        levels = [VOLUME_TABLE[v] / 32768.0 for v in self.vol_reg]
        periods = []
        for ch in range(3):
            reg = self.tone_reg[ch]
            # N <= 1 holds the output high instead of oscillating. That is
            # not a guard clause, it is the documented behaviour that makes
            # PSG sample playback possible at all.
            periods.append(float(reg) if reg > 1 else 0.0)

        noise_mode = self.noise_reg & 0x03
        noise_period = float(NOISE_RELOAD[noise_mode]) if noise_mode < 3 else \
            float(self.tone_reg[2] if self.tone_reg[2] > 0 else 1)

        for i in range(n_samples):
            total = 0.0
            for ch in range(3):
                period = periods[ch]
                if period <= 0.0:
                    total += levels[ch]           # DC high
                    continue
                total += levels[ch] * self._integrate_tone(ch, period,
                                                           ticks_per_sample)
            total += levels[3] * self._integrate_noise(noise_period,
                                                       ticks_per_sample)
            data[i] = total * 0.25                # /4, same headroom as psg.c
        return _audio.from_floats(data, 1)

    def _integrate_tone(self, ch: int, period: float, window: float) -> float:
        """Mean of the +/-1 square over one output sample."""
        remaining = window
        phase = self.tone_phase[ch]
        level = self.tone_output[ch]
        accumulated = 0.0
        while remaining > 0.0:
            if phase <= 0.0:
                phase = period
                level ^= 1
            span = phase if phase < remaining else remaining
            accumulated += span if level else -span
            phase -= span
            remaining -= span
        self.tone_phase[ch] = phase
        self.tone_output[ch] = level
        return accumulated / window

    def _integrate_noise(self, period: float, window: float) -> float:
        remaining = window
        phase = self.noise_phase
        accumulated = 0.0
        while remaining > 0.0:
            if phase <= 0.0:
                phase = period
                previous = self.noise_ff
                self.noise_ff ^= 1
                if previous == 0 and self.noise_ff == 1:
                    self._shift_lfsr()
            span = phase if phase < remaining else remaining
            accumulated += span if (self.lfsr & 1) else -span
            phase -= span
            remaining -= span
        self.noise_phase = phase
        return accumulated / window
